lib/services/service-utilities.ts

```diff
@@ -0,0 +1,204 @@
+ limitations under the License.
+ */
+
+import TypedError = require('typed-error');
+import * as Promise from 'bluebird';
+import * as bodyParser from 'body-parser';
+import * as express from 'express';
+import { Worker } from '../framework/worker';
+import { WorkerClient } from '../framework/worker-client';
+import { Logger, LogLevel } from '../utils/logger';
+import {
+	ServiceAPIHandle, ServiceEmitContext, ServiceEmitRequest, ServiceEmitResponse,
+	ServiceEmitter, ServiceListener, ServiceRegistration,
+} from './service-types';
+import {
+	UtilityServiceEvent, UtilityWorkerEvent
+} from './service-utilities-types';
+
+/**
+ * A utility class to handle a bunch of the repetitive work associated with being a ServiceListener and ServiceEmitter
+ *
+ * This class provides:
+ * - An express app, if asked for.
+ * - A logger.
+ * - An event registration and handling standard.
+ * - Checks for and extracts data from context.
+ * - A standard way of getting contextualised workers.
+ *
+ * In exchange you agree to:
+ * - provide a `emitData` method which takes a ServiceEmitContext and returns a promise.
+ * - provide a `verify` function which should check that incoming data isn't spoofed.
+ * - provide a `serviceName` getter, which because it is based on file path cannot be inherited away.
+ * - provide a `apiHandle` getter, which should return the underlying object that executes the requests.
+ * - enqueue your events with `context` and `event`, as per `UtilityServiceEvent`.
+ */
+export abstract class ServiceUtilities<T> extends WorkerClient<T> implements ServiceListener, ServiceEmitter {
+	/** A singleton express instance for all web-hook based services to share. */
+	private static _expressApp: express.Express;
+
+	/** A place to put output for debug and reference. */
+	private _logger = new Logger();
+
+	/** Store a list of actions to perform when particular actions happen */
+	private eventListeners: { [event: string]: ServiceRegistration[] } = {};
+
+	constructor() {
+		super();
+		this.logger.log(LogLevel.INFO, `---> '${this.serviceName}' constructing.`);
+	}
+
+	/**
+	 * Store an event of interest, so that the method gets triggered appropriately.
+	 * @param registration  Registration object with event trigger and other details.
+	 */
+	public registerEvent(registration: ServiceRegistration): void {
+		// Store each event registration in an object of arrays.
+		for (const event of registration.events) {
+			if (!this.eventListeners[event]) {
+				this.eventListeners[event] = [];
+			}
+			this.eventListeners[event].push(registration);
+		}
+	}
+
+	/**
+	 * Emit data to the service.
+	 * @param data  Service Emit Request to send, if relevant.
+	 * @returns     Details of the successful transmission from the service.
+	 */
+	public sendData(data: ServiceEmitRequest): Promise<ServiceEmitResponse> {
+		const context = data.contexts[this.serviceName] as ServiceEmitContext;
+		if (!context) {
+			return Promise.resolve({
```

Comment:
This is down to Client Bots being able to dispatch emit commands to multiple services at the same time using `dispatchToAllEmitters`. This means we can potentially end up in a situation where some dispatches have been successful, and some have failed. It's then up to the client to determine if any of the failures are actually a problem.